### PR TITLE
fix(vscode): Resolve Test Folder Addition, Naming Convention, and File Path Issues in Templates

### PR DESCRIPTION
**Pull request by samikay101**
# Pull Request
[Work Item for test folder fix](https://msazure.visualstudio.com/One/_workitems/edit/30079946)
[work Item for template updates](https://msazure.visualstudio.com/One/_workitems/edit/30079958)

## Requirement Checklist

* [ ] The commit message follows our guidelines
* [ ] Tests for the changes have been added (for bug fixes or features)
* [ ] Docs have been added or updated (for bug fixes or features)

## Type of Change

* [ ] Bug fix
* [ ] Feature
* [ ] Other

### 1. Adding the Tests Folder to the Workspace
   - **Issue**: The `Tests` folder was not automatically added to the workspace, causing difficulties for users in accessing and organizing generated test files.
   - **Action**: Updated the configuration to ensure the `Tests` folder is automatically created and added to the workspace when tests are generated, enhancing ease of access and organization.

### 2. Class and Method Naming Convention Update
   - **Issue**: The template configuration included `-` characters in class and method names, resulting in compilation errors due to incompatible naming.
   - **Action**: Adjusted naming conventions in all templates to replace `-` with `_`, ensuring compatibility and eliminating compilation errors. This update aligns the naming pattern with C# standards.

### 3. Standardizing Placement of `using` Statements
   - **Issue**: There were inconsistencies in the placement of `using` statements across templates, with some statements inside the namespace and others outside. Although best practices suggest placing `using` statements outside the namespace, some product-side configurations may differ.
   - **Action**: Finalized and updated templates to standardize the placement of `using` statements outside the namespace, bringing consistency and aligning with industry best practices.

### 4. Updating File Paths for Execution from the `bin` Folder
   - **Issue**: Current template paths did not account for execution from the `bin` directory, which led to runtime issues as files were not found in their expected locations.
   - **Action**: Modified file paths in class initializers to use relative paths suitable for execution from the `bin` folder. For example:
     ```csharp
     public orderprocesssuccessful()
     {
         this.workflowDefinitionPath = @"../../../../../LogicAppsDemo/la-sample-workflow/workflow.json";
         this.connectionsPath = @"../../../../../<logicappprojectname>/connections.json";
         this.parametersPath = @"../../../../../<logicappprojectname>/parameters.json";
         this.localSettingsPath = @"../../../../../<logicappprojectname>/local.settings.json";
         var mockData = JObject.Parse(File.ReadAllText(@"../../../<workflowname>/<unittestname>/<unittestname>.json"));
         this.triggerMock = mockData["triggerMocks"].ToObject<Dictionary<string, TriggerMock>>().FirstOrDefault().Value;
         this.actionMocks = mockData["actionMocks"].ToObject<Dictionary<string, ActionMock>>();
     }
     ```
     With these adjustments, files are located correctly during runtime.

## New Behavior

The updates introduce the following new behaviors:

1. **Automatic Addition of the `Tests` Folder**: The `Tests` folder is now automatically generated and added to the workspace for improved organization and accessibility of test files.
2. **Updated Naming Conventions**: All class and method names are configured to use `_` in place of `-`, preventing compilation errors and aligning with standard naming practices.
3. **Standardized Placement of `using` Statements**: Consistent placement of `using` statements outside the namespace across all templates.
4. **File Paths Configured for Execution from `bin` Folder**: Relative paths are configured to ensure all necessary files are accessible during runtime from the `bin` folder.

## Impact of Change

* [ ] **This is a breaking change.**

## Notable Limitation: Debug Session Dependency

A limitation exists when adding the `Tests` folder during an active debugging session, resulting in the following error:

> *"The following operation was blocked: Restarting extension host due to a workspace folder change. Reason: A debug session is still running."*

This error arises because the `vscode.workspace.updateWorkspaceFolders` API requires the extension host to restart, a process blocked by VS Code when a debug session is active to avoid disrupting debugging.

Despite this error message, the folder addition succeeds. The misleading error could cause confusion, so notifying users that the `Tests` folder is added successfully would be beneficial.

### Decoupling the API from Debug Session

The mock data generation API, essential for creating unit tests, is dependent on the debug session as it operates on a local port tied to this session. Restarting the extension host or ending the debug session would disrupt the API, making it unavailable for test generation. The dependency on the debug session, therefore, is currently unavoidable and may continue to cause this issue in future sessions.

## .csproj Updates and Observations

After attempting to generate an independent MSTest project, it was observed that several `ItemGroup` references were missing from the `.csproj` file. These references were added to determine if they would impact the detection of tests by the Test Explorer. The updates include the following:

```xml
<ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
    <PackageReference Include="MSTest.TestAdapter" Version="3.1.1" />
    <PackageReference Include="MSTest.TestFramework" Version="3.1.1" />
    <PackageReference Include="MSTest" Version="3.2.0" />
</ItemGroup>

### CODE REVIEW

**Comment by samikay101** [comment]
@nidhi-pathak

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile [hunk 1]
```diff
@@ -1,77 +1,128 @@
...
+            this.rootPath = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, @"..\..\..\..\.."));
+
+            // Build the full paths to the required JSON files.
+            this.workflowDefinitionPath = Path.Combine(rootPath, "<%= LogicAppName %>", "<%= WorkflowName %>", "workflow.json");
+            this.connectionsPath = Path.Combine(rootPath, "<%= LogicAppName %>", "connections.json");
+            this.parametersPath = Path.Combine(rootPath, "<%= LogicAppName %>", "parameters.json");
+            this.localSettingsPath = Path.Combine(rootPath, "<%= LogicAppName %>", "local.settings.json");
+
+            // Set path to the mock.json file.
+            var mockDataPath = Path.Combine(rootPath, "Tests", "<%= LogicAppName %>", "<%= WorkflowName %>", "<%= UnitTestName %>", "<%= UnitTestName %>-mock.json");
```
Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:68]: path of mock data should always be same folder as this test class file right? So file name should suffice here right?

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile [hunk 2]
```diff
@@ -1,77 +1,128 @@
...
+        public TriggerMock triggerMock;
+
+        /// <summary>
+        /// The mocked actions' outputs that are generated from a workflow run.
+        /// </summary>
+        public Dictionary<string, ActionMock> actionMocks;
+
+        /// <summary>
+        /// Initializes a new instance of the class named <see cref="<%= UnitTestName %>" />.
+        /// This constructor loads the necessary paths for the workflow, connections, parameters, and local settings JSON files.
```
Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:53]: i think we can remove these 2 lines

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile [hunk 3]
```diff
@@ -1,77 +1,128 @@
...
-            });
-
-            // Act 
-            var unitTestFlowRun = this.UnitTestExecutor.ExecuteWorkflow(this.WorkflowDefinition, this.Mocks, triggerMock).Result;
-
-            // Assert
-            Assert.IsNotNull(unitTestFlowRun);
-            Assert.AreEqual(FlowStatus.Succeeded, unitTestFlowRun.Properties.Status);
-            Assert.AreEqual("myvalue", unitTestFlowRun.Properties.Actions["Get_Rows_Receipt"].Outputs.Value<string>());
+            //Get the root path of the workspace
```
Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:58]: nit: space after start of comment.

and "Set the root path of the workspace.

Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:58]: Set the path for workflow related input files in the workspace.

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile [hunk 4]
```diff
@@ -1,77 +1,128 @@
...
+            this.rootPath = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, @"..\..\..\..\.."));
+
```
Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:60]: Remove these 2 lines

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile [hunk 5]
```diff
@@ -1,77 +1,128 @@
...
+            this.connectionsPath = Path.Combine(rootPath, "<%= LogicAppName %>", "connections.json");
+            this.parametersPath = Path.Combine(rootPath, "<%= LogicAppName %>", "parameters.json");
+            this.localSettingsPath = Path.Combine(rootPath, "<%= LogicAppName %>", "local.settings.json");
+
+            // Set path to the mock.json file.
+            var mockDataPath = Path.Combine(rootPath, "Tests", "<%= LogicAppName %>", "<%= WorkflowName %>", "<%= UnitTestName %>", "<%= UnitTestName %>-mock.json");
+
+            // Load the mock data from the corrected path.
+            var mockData = JObject.Parse(File.ReadAllText(mockDataPath));
+            this.triggerMock = mockData["triggerMocks"].ToObject<Dictionary<string, TriggerMock>>().FirstOrDefault().Value;
```
Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:72]: nit: FirstOrDefault()?.Value

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile [hunk 6]
```diff
@@ -1,77 +1,128 @@
...
-        public void ExecuteWorkflow_WithMockAction_ShouldSucceed_Dynamic()
+        public async Task <%= WorkflowName %>_<%= UnitTestName %>_ExecuteWorkflow_SUCCESS()
         {
-            // Arrange – define mocks either from a static source or as a lambda function
-            var triggerMock = this.Mocks["When_a_HTTP_request_is_received"];
-            this.Mocks["Get_Rows_Receipt"] = new OperationMock("Get_Rows_Receipt", FlowStatus.Succeeded, this.MockDynamicActionOutputCallback);
+            // PREPARE Mock
+            // var actionMocks = this.actionMocks;
+            // actionMocks["<actionName>"] = new ActionMock(
+            //     "<actionName>",
```
Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:86]: can we add named parameters here for better readablity

**Review by ccastrotrejo — APPROVED** [review]

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 2 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:53]: fixed

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 3 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:58]: fixed

Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:58]: fixed

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 5 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:72]: fixed

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 6 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:86]: fixed

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 1 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:68]: fixed

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 4 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile]: fixed

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile [hunk 7]
```diff
@@ -1,77 +1,129 @@
...
-            });
-
-            // Act 
-            var unitTestFlowRun = this.UnitTestExecutor.ExecuteWorkflow(this.WorkflowDefinition, this.Mocks, triggerMock).Result;
-
-            // Assert
-            Assert.IsNotNull(unitTestFlowRun);
-            Assert.AreEqual(FlowStatus.Succeeded, unitTestFlowRun.Properties.Status);
-            Assert.AreEqual("myvalue", unitTestFlowRun.Properties.Actions["Get_Rows_Receipt"].Outputs.Value<string>());
+            this.rootPath = Path.GetFullPath(path: Path.Combine(basePath: AppContext.BaseDirectory, relativePath: @"..\..\..\..\.."));
```
Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:56]: duplicate code?

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile [hunk 8]
```diff
@@ -1,77 +1,129 @@
...
-            var unitTestFlowRun = this.UnitTestExecutor.ExecuteWorkflow(this.WorkflowDefinition, this.Mocks, triggerMock).Result;
+            // ACT
+            // Create an instance of UnitTestExecutor, and run the workflow with the mock data.
+            var executor = new UnitTestExecutor(
+                workflowDefinitionPath: this.workflowDefinitionPath, 
+                connectionsPath: this.connectionsPath, 
+                parametersPath: this.parametersPath, 
+                localSettingsPath: this.localSettingsPath
+            );            
+
```
Comment (nidhi-pathak) [inline apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile:92]: nit: omit newline

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 1 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile]: @nidhi-pathak Based on the PR feedback, it appears the folder could not be located at the specified path. To keep us on track, I’ll be reverting to the most recent working version for now.

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 8 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile]: fixed

File: apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile (hunk 7 above)
Comment (samikay101) [reply apps/vs-code-designer/src/assets/UnitTestTemplates/TestClassFile]: fixed

**Pull request merged**